service/resource/cloudformation/adaptor.go

```diff
@@ -0,0 +1,85 @@
+package cloudformation
+
+import (
+	"github.com/giantswarm/awstpr"
+	"github.com/giantswarm/microerror"
+	"golang.org/x/sync/errgroup"
+
+	awsutil "github.com/giantswarm/aws-operator/client/aws"
+)
+
+type hydrater func(awstpr.CustomObject, awsutil.Clients) error
+
+type adaptor struct {
+	ASGType string
+
+	// worker launch configuration
```

Comment:
I like the adaptor structure but I wonder how it will scale as we add more resources. Some attributes like SecurityGroupID will be needed multiple times and will be different for master and workers.

So I'm worried about collisions. Will we have multiple adaptors or can we nest structs within this adaptor struct?

Reply:
Totally, I started having separate types for each component, and then a main adaptor type that embedded them anonymously, so that all the fields were available when doing the template execution.

IMO this approach is better than having multiple adaptors, that would lead us to execute separately each template and then merge the resulting buffers, i find it more natural to have this last stage in a single step (one single template execution with one single data struct), let me know WDYT.

Reply:
Yes I think having a single adaptor with per component types is good. I also prefer a single template execution.

Reply:
Done, now we have separate types for each template, they are embedded as anonymous members in the main adapter so that the fields are accessible from the included templates and, more importantly, we let the compiler help us with ambiguous fields. 

About those common fields, I think we can promote them to the top level adapter type, so that we create them once and can be used in all the embedded templates, WDYT?